zrml/court/src/lib.rs

```diff
@@ -16,266 +16,1726 @@
+                }
+            } else {
+                pool = remove_weakest_if_full(pool)?;
+            }
+
+            let (active_lock, consumed_stake, joined_at) = (active_lock, consumed_stake, joined_at);
+
+            match pool.binary_search_by_key(&(amount, who), |pool_item| {
+                (pool_item.stake, &pool_item.court_participant)
+            }) {
+                Ok(_) => {
+                    debug_assert!(
+                        false,
+                        "This should never happen, because we are removing the court participant \
+                         above."
+                    );
+                    return Err(Error::<T>::CourtParticipantTwiceInPool.into());
+                }
+                Err(i) => pool
+                    .try_insert(
+                        i,
+                        CourtPoolItem {
+                            stake: amount,
+                            court_participant: who.clone(),
+                            consumed_stake,
+                            joined_at,
+                        },
+                    )
+                    .map_err(|_| {
+                        debug_assert!(
+                            false,
+                            "This should never happen, because we are removing the lowest staked \
+                             court participant above."
+                        );
+                        Error::<T>::MaxCourtParticipantsReached
+                    })?,
+            };
+
+            T::Currency::set_lock(T::LockId::get(), who, amount, WithdrawReasons::all());
+
+            let pool_len = pool.len() as u32;
+            CourtPool::<T>::put(pool);
+
+            let p_info = CourtParticipantInfoOf::<T> {
+                stake: amount,
+                active_lock,
+                prepare_exit_at: None,
+                delegations,
+            };
+            <Participants<T>>::insert(who, p_info);
+
+            Ok(pool_len)
+        }
+
+        // Handle the external incentivisation of the court system.
+        pub(crate) fn handle_inflation(now: T::BlockNumber) -> Weight {
+            let inflation_period = T::InflationPeriod::get();
+            if (now % inflation_period).is_zero() {
+                let yearly_inflation_rate = <YearlyInflation<T>>::get();
+                let yearly_inflation_amount = yearly_inflation_rate * T::Currency::total_issuance();
+                let blocks_per_year = T::BlocksPerYear::get()
+                    .saturated_into::<u128>()
+                    .saturated_into::<BalanceOf<T>>();
+                debug_assert!(!T::BlocksPerYear::get().is_zero());
+                let issue_per_block = yearly_inflation_amount / blocks_per_year.max(One::one());
+
+                let inflation_period_mint = issue_per_block.saturating_mul(
+                    inflation_period.saturated_into::<u128>().saturated_into::<BalanceOf<T>>(),
+                );
```

Comment:
I calculated the actual values and the mint exceeds 0.01% for one period.

The mint for one 30 day period is currently `1723632215616000` and 0.01% of `total_issuance` is `1049261111270354576 * 0.0001 = 104926111127035`.

Reply:
If I want to increase the total supply 2% in a year and assume the inflation period is month, you get `0.02 / 12 = 0.00166667`. So it is `0.166667 %` each month. So the `inflation_period_mint` heavily depends on the `InflationPeriod`.

Reply:
I could say, the `inflation_period_mint` should never exceed the 0.2% of the total supply, but this heavily depends on the inflation period block number. So whenever we change the inflation period in production it could lead to a non-mint bug (if we forget to adjust this "abort + debug_assert".

Reply:
I think a simple trivial safre-guard is sufficient, for example `inflation_period_mint <= <YearlyInflation<T>>::get()`. Should that guard trigger, we definitely will have to abort execution.

Reply:
It would also be nice to have log warnings when the value is suspiciously high. Maybe not the best practice but we could specify a constant here to `0.5%`, which seems fair enough.

Reply:
e471c1e267bd8fd8d2d54cd3c70e6f4794f057f1

Reply:
4dd09310c5aaf85af80f21e48f61ad70a7f96f45